lib/scala/logging-service-logback/src/main/java/org/enso/logging/service/logback/telemetry/TelemetryAppender.java

```diff
@@ -0,0 +1,270 @@
+ * {@link TelemetryAppender#append(ILoggingEvent)} method are received from a socket and
+ * deserialized by the logback framework. Thus, the {@link ILoggingEvent#getArgumentArray() log
+ * event arguments} are most likely strings.
+ */
+public final class TelemetryAppender extends AppenderBase<ILoggingEvent> {
+  private static final String CREDENTIALS_FILE_ENV = "ENSO_CLOUD_CREDENTIALS_FILE";
+
+  /**
+   * We still want to limit the batch size to some reasonable number - sending too many logs in one
+   * request could also be problematic.
+   */
+  private static final int MAX_BATCH_SIZE = 100;
+
+  private static final int MAX_RETRIES = 5;
+  private static final Logger LOGGER = LoggerFactory.getLogger(TelemetryAppender.class.getName());
+  private static TelemetryAppender instance;
+
+  private Credentials credentials;
+  private final LogJobsQueue logQueue = new LogJobsQueue();
+  private final ThreadPoolExecutor backgroundThreadService;
+  private final URI endpoint;
+
+  private HttpClient httpClient;
+
+  /**
+   * Set to true once an error is encountered when sending a request. In such case, it is most
+   * probably that no further requests will be successful. This flag is used to terminate the
+   * background thread.
+   */
+  private boolean requestSendingFailure;
+
+  private TelemetryAppender(ThreadPoolExecutor backgroundThreadService, URI endpoint) {
+    this.backgroundThreadService = backgroundThreadService;
+    this.endpoint = endpoint;
+  }
+
+  public static TelemetryAppender getInstance() {
+    if (instance == null) {
+      instance = create();
+    }
+    return instance;
+  }
+
+  private static TelemetryAppender create() {
+    // We set-up a thread 'pool' that will contain at most one thread.
+    // If the thread is idle for 60 seconds, it will be shut down.
+    var executor = new ThreadPoolExecutor(0, 1, 60L, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+
+    var endpoint = URI.create(getCloudLogsAPIEndpoint());
+    return new TelemetryAppender(executor, endpoint);
+  }
+
+  @Override
+  protected void append(ILoggingEvent eventObject) {
+    enqueueJob(eventObject);
+  }
+
+  private static Path credentialsFile() {
+    var env = System.getenv(CREDENTIALS_FILE_ENV);
+    if (env != null) {
+      return Path.of(env);
+    }
+    var home = Path.of(System.getProperty("user.home"));
+    var credentials = home.resolve(".enso").resolve("credentials");
+    return credentials;
+  }
+
+  private static Credentials parseCredentials(Path file) throws IOException {
+    assert file.toFile().exists();
+    var objectMapper = new ObjectMapper();
+    return objectMapper.readValue(file.toFile(), Credentials.class);
+  }
+
+  private static String getCloudLogsAPIEndpoint() {
+    var envUri = System.getenv("ENSO_CLOUD_API_URI");
+    var effectiveUri =
+        envUri == null ? "https://7aqkn3tnbc.execute-api.eu-west-1.amazonaws.com/" : envUri;
```

Comment:
remove hardcoded path

Reply:
This was copied from [org.enso.base.enso_cloud.CloudAPI#getAPIRootURI](https://github.com/enso-org/enso/blob/8c900d5eb4eca019b7afad3e46d27df7e006cddf/std-bits/base/src/main/java/org/enso/base/enso_cloud/CloudAPI.java#L7-L18). There is no other simple way to get URL to the production cloud. Any suggestions?

Reply:
OK, fair point. That doesn't look like the right place for that either. On top of that I don't want that uri littered in various locations. If it is the default prod one, then it should be provided externally. I assume that this was added there for convenience, for pure cli execution, right @radeusgd?

Reply:
> If it is the default prod one, then it should be provided externally

Yes, it should. Added as a task to https://github.com/enso-org/enso/issues/12564

Reply:
So much TODO. So little time.